### DISCUSSION TITLE
How to set limits for a parameter?

### DISCUSSION THREAD

**Discussion by Ricardo-DO** [nlmixr2/rxode2#514] [General]
Hello everyone,

I was trying to use RxODE (by now, I am planning to move to rxode2 but I having problems with the installation) to replicate a code I was using in Berkeley Madonna.

It is a dissolution model that I am planning to implement in a compartmental model.
The code is the following
```r
#############################################
# Define model
ode <- "
   d/dt(Xv) = z*X0*(((X0-Xv)/X0)^(1/3))*(Cs-(Xv/V));
"

# Compile model
mod1 <- RxODE(model = ode, modName = "mod1")

# Define system parameters
params <- 
   c(z  = 0.008,	
     V  = 250, 		
     X0 =   20 ,	
     Cs = 0.1)		

inits <- 0 

# Initialize event table
ev <- eventTable()

# Specify sampling
ev$add.sampling(0:600) #in seconds

# Simulate
x <- mod1$run(params, ev, inits)
#######################################
````
The code runs ok in those conditions, but if I change Cs to 1.6, Xv becomes larger than X0 after 100 s and I only get NaN as output.

Is there a way to specify an ifelse function like
Xv<-ifelse(Xv<X0, Xv,X0) ?

The problem is that, in those conditions X0-Xv will be <0 
I tried to create another parameter, let's call it Xs = X0-Xv, but it says that Xs is needed to solve the system.

I tried creating the ifelse statement like
if (Xv<X0){
Xv=Xv} else{
 Xv=X0}

But that doesn't seem to be the answer.

In Berkeley Madonna thee is a function called LIMIT which does that, is there something similar in RxODE?

If you have any thoughts it will be much appreciated.

Many thanks,
Ricardo

**Comment by mattfidler** [comment]
Hi @Ricardo-DO,

Thank you for reaching out.

I modified your original post so the code in your comment is interpreted as R code so it looks a little nicer and the `*` signs don't disappear.

You suggested using a `ifelse()` expression, which should work.  For my example, I have recast your model in `rxode2` model functions to use model piping to change initial conditions and model lines. 

Depending on the version of `RxODE` the `ifelse()` may or may not be supported.  

``` r
library(rxode2)
#> Warning: package 'rxode2' was built under R version 4.2.3
#> rxode2 2.0.13 using 4 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`

f <- function() {
  ini({
    z = 0.008
    V = 250
    X0 = 20
    Cs = 0.1
  })
  model({
    d/dt(Xv) = z*X0*(((X0-Xv)/X0)^(1/3))*(Cs-(Xv/V))
  })
}

ev <- eventTable()

ev$add.sampling(0:600) #in seconds

x <- rxSolve(f, ev)
# runs fine
print(x)
#> ── Solved rxode2 object ──
#> ── Parameters ($params): ──
#>       z       V      X0      Cs 
#>   0.008 250.000  20.000   0.100 
#> ── Initial Conditions ($inits): ──
#> Xv 
#>  0 
#> ── First part of data (object): ──
#> # A tibble: 601 × 2
#>    time     Xv
#>   <dbl>  <dbl>
#> 1     0 0     
#> 2     1 0.0160
#> 3     2 0.0320
#> 4     3 0.0479
#> 5     4 0.0639
#> 6     5 0.0798
#> # ℹ 595 more rows

# update to Cs = 1.6
f2 <- f |> ini(Cs=1.6)
#> ℹ change initial estimate of `Cs` to `1.6`

x <- rxSolve(f2, ev)
#  runs fine, but some values are NaN
summary(x)
#> ── Summary of Solved rxode2 object ──
#> ── Model ($model): ──
#> rxode2({
#>     param(z, V, X0, Cs)
#>     d/dt(Xv) = z * X0 * (((X0 - Xv)/X0)^(1/3)) * (Cs - (Xv/V))
#> }) 
#> ── Parameters (x$params): ──
#>       z       V      X0      Cs 
#>   0.008 250.000  20.000   1.600 
#> ── Initial Conditions (x$inits): ──
#> Xv 
#>  0 
#> ── Summary of data (object): ──
#>       time           Xv       
#>  Min.   :  0   Min.   : 0.00  
#>  1st Qu.:150   1st Qu.: 7.05  
#>  Median :300   Median :12.91  
#>  Mean   :300   Mean   :11.95  
#>  3rd Qu.:450   3rd Qu.:17.40  
#>  Max.   :600   Max.   :19.96  
#>                NA's   :481

# replace the line in the mdel with ifelse()

f3 <- f2 |> 
  model({d/dt(Xv) <-ifelse(Xv<X0, z*X0*(((X0-Xv)/X0)^(1/3))*(Cs-(Xv/V)),X0)})

print(f3) 
#>  ── rxode2-based free-form 1-cmt ODE model ────────────────────────────────────── 
#>  ── Initalization: ──  
#> Fixed Effects ($theta): 
#>       z       V      X0      Cs 
#>   0.008 250.000  20.000   1.600 
#> 
#> States ($state or $stateDf): 
#>   Compartment Number Compartment Name
#> 1                  1               Xv
#>  ── Model (Normalized Syntax): ── 
#> function() {
#>     ini({
#>         z <- 0.008
#>         V <- 250
#>         X0 <- 20
#>         Cs <- 1.6
#>     })
#>     model({
#>         d/dt(Xv) <- ifelse(Xv < X0, z * X0 * (((X0 - Xv)/X0)^(1/3)) * 
#>             (Cs - (Xv/V)), X0)
#>     })
#> }

x <- rxSolve(f3, ev)

#  No values are NaN by using ifelse
summary(x)
#> ── Summary of Solved rxode2 object ──
#> ── Model ($model): ──
#> rxode2({
#>     param(z, V, X0, Cs)
#>     d/dt(Xv) = ifelse(Xv < X0, z * X0 * (((X0 - Xv)/X0)^(1/3)) * 
#>         (Cs - (Xv/V)), X0)
#> }) 
#> ── Parameters (x$params): ──
#>       z       V      X0      Cs 
#>   0.008 250.000  20.000   1.600 
#> ── Initial Conditions (x$inits): ──
#> Xv 
#>  0 
#> ── Summary of data (object): ──
#>       time           Xv        
#>  Min.   :  0   Min.   :   0.0  
#>  1st Qu.:150   1st Qu.: 603.1  
#>  Median :300   Median :3603.1  
#>  Mean   :300   Mean   :3846.5  
#>  3rd Qu.:450   3rd Qu.:6603.1  
#>  Max.   :600   Max.   :9603.1
```

<sup>Created on 2023-04-29 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>


Note that the way you use `RxODE()` will still work with `rxode2()`, if you prefer to keep that style of coding

``` r
library(rxode2)
#> Warning: package 'rxode2' was built under R version 4.2.3
#> rxode2 2.0.13 using 4 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`

#############################################
# Define model
ode <- "
   d/dt(Xv) = z*X0*(((X0-Xv)/X0)^(1/3))*(Cs-(Xv/V));
"

# Compile model
mod1 <- RxODE(model = ode, modName = "mod1")

# Define system parameters
params <- 
  c(z  = 0.008, 
    V  = 250,       
    X0 =   20 , 
    Cs = 0.1)       

inits <- 0 

# Initialize event table
ev <- eventTable()

# Specify sampling
ev$add.sampling(0:600) #in seconds

# Simulate
x <- mod1$run(params, ev, inits)
#> Warning: Assumed order of inputs: Xv
#####################################
```

<sup>Created on 2023-04-29 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Reply by Ricardo-DO** [reply]
Hi @mattfidler 

Many thanks for taking the time to reply to my question, really appreciate it!

Thanks for tidying up the code, it's neater now.
I ran the code with the ifelse and while is not giving any NaN values, there is something going on when Xv < X0 is not TRUE.

I get the same summary:

── Parameters (x$params): ──
      z       V      X0      Cs 
  0.008 250.000  20.000   1.600 
── Initial Conditions (x$inits): ──
Xv 
 0 
── Summary of data (object): ──
      time           Xv        
 Min.   :  0   Min.   :   0.0  
 1st Qu.:150   1st Qu.: 603.1  
 Median :300   Median :3603.1  
 Mean   :300   Mean   :3846.5  
 3rd Qu.:450   3rd Qu.:6603.1  
 Max.   :600   Max.   :9603.1  

As we want that when Xv >= X0, the equation changes to X0, the max value should be X0, which in this case is 20.

The plot looks like this

![image](https://user-images.githubusercontent.com/59565610/235371076-0c0a80da-f48a-4062-bbc9-93dc1b508df9.png)

If I ran this similar code in Berkeley Madonna 

METHOD RK4
STARTTIME = 0
STOPTIME=600
DT = 1

init xdv = 0

x0 = 20; 
zf = 0.008; 
Vv = 250; 
Cs = 1.6; 

d/dt (xdv) = zf*x0 * (((x0-xdv)/x0)^(1/3)) * (Cs - (xdv/Vv))
LIMIT xdv <= x0


I get this plot

![image](https://user-images.githubusercontent.com/59565610/235371414-a02e0b8c-8fe4-4fe4-bcff-9b3200df96b4.png)


Do you think that there is something that it wasn't captured?

Any advice would be greatly appreciated.

Many thanks,
Ricardo

**Comment by Ricardo-DO** [comment]
Sorry for the 2 replies.

I kept trying things, and I made this change to the code (which I acknwledge is not the best solution, but the result is similar to what I'm looking for)

If in the ifelse statement we keep the same equation but instead of X0-Xv we directly write 0  as the else condition, it kind of works

library(rxode2)

f <- function() {
  ini({
    z = 0.008
    V = 250
    X0 = 20
    Cs = 1.6
  })
  model({
     d/dt(Xv) <- ifelse(Xv < X0, z * X0 * (((X0 - Xv)/X0)^(1/3)) * (Cs - (Xv/V)), z * X0 * (((0)/X0)^(1/3)) * (Cs - (Xv/V)))
  })
}

ev <- eventTable()

ev$add.sampling(0:3600) #in seconds

x <- rxSolve(f, ev)


![image](https://user-images.githubusercontent.com/59565610/235372448-73c0421d-b7b6-440f-9f93-f1fbb6eb59bf.png)


── Summary of Solved rxode2 object ──
── Model ($model): ──
rxode2({
    param(z, V, X0, Cs)
    d/dt(Xv) = ifelse(Xv < X0, z * X0 * (((X0 - Xv)/X0)^(1/3)) * 
        (Cs - (Xv/V)), z * X0 * (((0)/X0)^(1/3)) * (Cs - (Xv/V)))
}) 
── Parameters (x$params): ──
      z       V      X0      Cs 
  0.008 250.000  20.000   1.600 
── Initial Conditions (x$inits): ──
Xv 
 0 
── Summary of data (object): ──
      time            Xv       
 Min.   :   0   Min.   : 0.00  
 1st Qu.: 900   1st Qu.:20.47  
 Median :1800   Median :21.00  
 Mean   :1800   Mean   :20.74  
 3rd Qu.:2700   3rd Qu.:21.54  
 Max.   :3600   Max.   :22.08  

We can see that now, the max should be 20 and it is 22.08, but it's closer to 20 in this way.

Thanks,
Ricardo

**Reply by mattfidler** [reply]
This is probably what you are looking for:

``` r
library(rxode2)
#> rxode2 2.0.13.9000 using 8 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`

f <- function() {
  ini({
    z = 0.008
    V = 250
    X0 = 20
    Cs = 1.6
  })
  model({
     d/dt(Xv) <- ifelse(Xv <= X0, z * X0 * (((X0 - Xv)/X0)^(1/3)) * (Cs - (Xv/V)), 0)
  })
}

ev <- eventTable()

ev$add.sampling(0:3600) #in seconds

x <- rxSolve(f, ev)

library(ggplot2)

plot(x) + xlim(0, 600)
#> Scale for x is already present.
#> Adding another scale for x, which will replace the existing scale.
#> Warning: Removed 3000 rows containing missing values (`geom_line()`).
```

![](https://i.imgur.com/hdl0yID.png)

<sup>Created on 2023-05-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Reply by Ricardo-DO** [reply]
That's great @mattfidler!

That is what I was lloking for!
Many thanks!

Best,
Ricardo
